File: frontend/src/constants/supplierTaxonomy.ts
/**
 * Must match backend `catalog/supplier_taxonomy.py` (validated on API).
 * For runtime fetch use GET /suppliers/taxonomy if you need server as source of truth.
 */
export const SUPPLIER_COUNTRIES: { value: string; label: string; isEu: boolean }[] = [
  { value: "Polska", label: "Polska", isEu: true },
  { value: "Niemcy", label: "Niemcy", isEu: true },
  { value: "Francja", label: "Francja", isEu: true },
  { value: "Czechy", label: "Czechy", isEu: true },
  { value: "Anglia", label: "Anglia", isEu: false },
  { value: "Hiszpania", label: "Hiszpania", isEu: true },
  { value: "Chiny", label: "Chiny", isEu: false },
];

export const SUPPLIER_COUNTRY_VALUES = new Set(SUPPLIER_COUNTRIES.map((c) => c.value));

export const SUPPLIER_CURRENCIES = ["EUR", "DOL", "CNY", "CZK", "PLN"] as const;
export type SupplierCurrencyCode = (typeof SUPPLIER_CURRENCIES)[number];

export function isEuCountryLabel(country: string | null | undefined): boolean | null {
  if (country == null || !String(country).trim()) return null;
  const row = SUPPLIER_COUNTRIES.find((c) => c.value === country.trim());
  return row ? row.isEu : null;
}

File: backend/catalog/supplier_taxonomy.py
"""
Canonical supplier field values (country, currency).

Extension point — product↔supplier linkage
-----------------------------------------
Today, assortment uses ``Product.default_supplier_id`` only. When a many-to-many
``supplier_products`` (or similar) table exists, centralize “which products belong
to supplier S” in ``supplier_product_linkage.py`` and keep this module for
enumerated supplier attributes only.
"""

from __future__ import annotations

from typing import FrozenSet, List, Optional, Tuple

# (display label stored in DB, is EU member — for future VAT rules)
SUPPLIER_COUNTRY_CHOICES: Tuple[Tuple[str, bool], ...] = (
    ("Polska", True),
    ("Niemcy", True),
    ("Francja", True),
    ("Czechy", True),
    ("Anglia", False),
    ("Hiszpania", True),
    ("Chiny", False),
)

ALLOWED_SUPPLIER_COUNTRIES: FrozenSet[str] = frozenset(c for c, _ in SUPPLIER_COUNTRY_CHOICES)

_COUNTRY_EU: dict[str, bool] = {c: eu for c, eu in SUPPLIER_COUNTRY_CHOICES}

ALLOWED_SUPPLIER_CURRENCIES: FrozenSet[str] = frozenset({"EUR", "DOL", "CNY", "CZK", "PLN"})


def country_is_eu(country: Optional[str]) -> Optional[bool]:
    """None if country empty or unknown (legacy row not in catalog)."""
    if country is None:
        return None
    t = str(country).strip()
    if not t:
        return None
    return _COUNTRY_EU.get(t)


def validate_supplier_country(value: Optional[str]) -> Optional[str]:
    if value is None:
        return None
    t = str(value).strip()
    if not t:
        return None
    if t not in ALLOWED_SUPPLIER_COUNTRIES:
        raise ValueError(
            f"Invalid country: must be one of {sorted(ALLOWED_SUPPLIER_COUNTRIES)}",
        )
    return t


def validate_supplier_currency(value: Optional[str]) -> Optional[str]:
    if value is None:
        return None
    t = str(value).strip().upper()
    if not t:
        return None
    if t not in ALLOWED_SUPPLIER_CURRENCIES:
        raise ValueError(
            f"Invalid currency: must be one of {sorted(ALLOWED_SUPPLIER_CURRENCIES)}",
        )
    return t


def list_country_choices() -> List[dict]:
    """For optional meta API / docs."""
    return [{"value": c, "label": c, "is_eu": eu} for c, eu in SUPPLIER_COUNTRY_CHOICES]


def list_currency_choices() -> List[str]:
    return sorted(ALLOWED_SUPPLIER_CURRENCIES)
